test/helpers/ProposalTest.sol

```diff
@@ -0,0 +1,334 @@
+      string memory _description
+    )
+  {
+    (_targets, _values, _calldata, _description) =
+      _buildTokenSendProposal(_token, _amount, _receiver);
+
+    // Submit the new proposal
+    vm.prank(PROPOSER);
+    _newProposalId = governorBravo.propose(_targets, _values, _calldata, _description);
+
+    // Ensure proposal is in the expected state
+    IGovernor.ProposalState _state = governorBravo.state(_newProposalId);
+    assertEq(_state, IGovernor.ProposalState.Pending);
+  }
+
+  // Take a proposal through its full lifecycle, from proposing it, to voting on
+  // it, to queuing it, to executing it (if relevant) via GovernorBravo.
+  function _queueAndVoteAndExecuteProposalWithBravoGovernor(
+    address[] memory _targets,
+    uint256[] memory _values,
+    bytes[] memory _calldatas,
+    string memory _description,
+    uint8 _voteType
+  ) internal {
+    // Submit the new proposal
+    vm.prank(PROPOSER);
+    uint256 _newProposalId = governorBravo.propose(_targets, _values, _calldatas, _description);
+
+    // Ensure proposal is Pending.
+    IGovernor.ProposalState _state = governorBravo.state(_newProposalId);
+    assertEq(_state, IGovernor.ProposalState.Pending);
+
+    _jumpToActiveProposal(_newProposalId);
+
+    // Have all delegates cast their weight with the specified support type.
+    _delegatesCastVoteOnBravoGovernor(_newProposalId, _voteType);
+
+    _jumpToVotingComplete(_newProposalId);
+
+    _state = governorBravo.state(_newProposalId);
+    if (_voteType == AGAINST || _voteType == ABSTAIN) {
+      // The proposal should have failed.
+      assertEq(_state, IGovernor.ProposalState.Defeated);
+
+      // Attempt to queue the proposal.
+      vm.expectRevert("Governor: proposal not successful");
+      governorBravo.queue(_targets, _values, _calldatas, keccak256(bytes(_description)));
+
+      _jumpPastProposalEta(_newProposalId);
+
+      // Attempt to execute the proposal.
+      vm.expectRevert("Governor: proposal not successful");
+      governorBravo.execute(_targets, _values, _calldatas, keccak256(bytes(_description)));
+
+      // Exit this function, there's nothing left to test.
+      return;
+    }
+
+    // The voteType was FOR. Ensure the proposal has succeeded.
+    assertEq(_state, IGovernor.ProposalState.Succeeded);
+
+    // Queue the proposal
+    governorBravo.queue(_targets, _values, _calldatas, keccak256(bytes(_description)));
+
+    // Ensure the proposal is queued
+    _state = governorBravo.state(_newProposalId);
+    assertEq(_state, IGovernor.ProposalState.Queued);
+
+    _jumpPastProposalEta(_newProposalId);
+
+    // Execute the proposal
+    governorBravo.execute(_targets, _values, _calldatas, keccak256(bytes(_description)));
+
+    // Ensure the proposal is executed
+    _state = governorBravo.state(_newProposalId);
+    assertEq(_state, IGovernor.ProposalState.Executed);
+  }
+
+  function assertEq(IGovernor.ProposalState _actual, IGovernor.ProposalState _expected) internal {
```

Comment:
Were these methods copied from the PoolTogether upgrade or are there some modifications.

Reply:
Copied from PoolTogether.. not modified.